nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp

```diff
@@ -0,0 +1,268 @@
+    return;
+  }
+
+  // Set base_ and multiplier_
+  base_ = msg->base;
+  multiplier_ = msg->multiplier;
+  // Set topic name to receive filter mask from
+  mask_topic_ = msg->filter_mask_topic;
+
+  // Setting new filter mask subscriber
+  RCLCPP_INFO(
+    logger_,
+    "BinaryFilter: Subscribing to \"%s\" topic for filter mask...",
+    mask_topic_.c_str());
+  mask_sub_ = node->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    mask_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&BinaryFilter::maskCallback, this, std::placeholders::_1));
+}
+
+void BinaryFilter::maskCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    RCLCPP_INFO(
+      logger_,
+      "BinaryFilter: Received filter mask from %s topic.", mask_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "BinaryFilter: New filter mask arrived from %s topic. Updating old filter mask.",
+      mask_topic_.c_str());
+    filter_mask_.reset();
+  }
+
+  filter_mask_ = msg;
+  mask_frame_ = msg->header.frame_id;
+}
+
+void BinaryFilter::process(
+  nav2_costmap_2d::Costmap2D & /*master_grid*/,
+  int /*min_i*/, int /*min_j*/, int /*max_i*/, int /*max_j*/,
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    // Show warning message every 2 seconds to not litter an output
+    RCLCPP_WARN_THROTTLE(
+      logger_, *(clock_), 2000,
+      "BinaryFilter: Filter mask was not received");
+    return;
+  }
+
+  geometry_msgs::msg::Pose2D mask_pose;  // robot coordinates in mask frame
+
+  // Transforming robot pose from current layer frame to mask frame
+  if (!transformPose(global_frame_, pose, mask_frame_, mask_pose)) {
+    return;
+  }
+
+  // Converting mask_pose robot position to filter_mask_ indexes (mask_robot_i, mask_robot_j)
+  unsigned int mask_robot_i, mask_robot_j;
+  if (!worldToMask(filter_mask_, mask_pose.x, mask_pose.y, mask_robot_i, mask_robot_j)) {
+    // Robot went out of mask range. Set "false" state by-default
+    RCLCPP_WARN(
+      logger_,
+      "BinaryFilter: Robot is outside of filter mask. Resetting binary state to default.");
+    changeState(default_state_);
+    return;
+  }
+
+  // Getting filter_mask data from cell where the robot placed
+  int8_t mask_data = getMaskData(filter_mask_, mask_robot_i, mask_robot_j);
+  if (mask_data == nav2_util::OCC_GRID_UNKNOWN) {
+    // Corresponding filter mask cell is unknown.
+    // Warn and do nothing.
+    RCLCPP_WARN_THROTTLE(
+      logger_, *(clock_), 2000,
+      "BinaryFilter: Filter mask [%i, %i] data is unknown. Do nothing.",
+      mask_robot_i, mask_robot_j);
+    return;
+  }
+  // Check and flip binary state, if necessary
+  if (base_ + mask_data * multiplier_ > flip_threshold_) {
+    if (binary_state_ == default_state_) {
```

Comment:
I don't understand the comparison with `default_state_`. I don't think there should be any checking happening here beyond just that binary_state_ is true or false in order to either change the state or not based on if the data is above the threshold e.g.

```
if (!binary_state_) {changeState(false);}
```

Reply:
Or is your intention that above the threshold moves from the default state to the non-default state? I suppose that would be more general.